### ISSUE TITLE
Cannot read ONNX model, error in reshape_layer.cpp, function 'computeShapeByReshapeMask'

### ISSUE THREAD

**Issue by eugeneRover** [opencv/opencv#15576]
#### System information (version)
- OpenCV => 4.1.1 (master branch, commit a74fe2ec01d9218d06cb7675af633fc3f409a6a2)
- Operating System / Platform => Debian Linux 64 Bit
- Compiler => gcc version 6.3.0 20170516 (Debian 6.3.0-18+deb9u1)

##### Detailed description

Error while trying to read pretrained ONNX DNN model file:

> cv2.error: OpenCV(4.1.2-pre) /media/sdb2/opencv/modules/dnn/src/layers/reshape_layer.cpp:125: error: (-5:Bad argument) Copy dim[3] (which has zero size) is out of the source shape bounds in function 'computeShapeByReshapeMask'

##### Steps to reproduce

```.python
net = cv.dnn.readNetFromONNX('/media/sdb2/temp/mask_rcnn_R_50_FPN_1x.onnx')
```

Model file is available online at the [page](https://github.com/onnx/models/tree/master/vision/object_detection_segmentation/mask-rcnn)  ([direct link](https://onnxzoo.blob.core.windows.net/models/opset_10/mask_rcnn/mask_rcnn_R_50_FPN_1x.onnx) )

I also tried model from [other page](https://github.com/onnx/models/tree/master/vision/object_detection_segmentation/faster-rcnn) with the same error

**Comment by bencha30** [comment]
same problem here..

trained model: ssd_mobilenet_v2_coco
Ubuntu 16.04
using tensorflow object detection API (train.py)
exported the graph (.pb file and .pbtx using tf_text_graph_ssd.py)

cvNet = cv2.dnn.readNetfromTensorflow('./ssd_mobilenet_v2_coco_trained/frozen_inference_graph.pb', './ssd_mobilenet_v2_coco_trained/graph1.pbtxt')

any suggestion on how to solve this ?

error:
 cvOut = cvNet.forward()
cv2.error: OpenCV(4.1.1) C:\projects\opencv-python\opencv\modules\dnn\src\layers\reshape_layer.cpp:146: error: (-1:Backtrace) Can't infer a dim denoted by -1 in function 'cv::dnn::computeShapeByReshapeMask'

**Comment by jf99** [comment]
Still the same error with OpenCV 4.2.0.

**Comment by zacario-li** [comment]
> same problem here..
> 
> trained model: ssd_mobilenet_v2_coco
> Ubuntu 16.04
> using tensorflow object detection API (train.py)
> exported the graph (.pb file and .pbtx using tf_text_graph_ssd.py)
> 
> cvNet = cv2.dnn.readNetfromTensorflow('./ssd_mobilenet_v2_coco_trained/frozen_inference_graph.pb', './ssd_mobilenet_v2_coco_trained/graph1.pbtxt')
> 
> any suggestion on how to solve this ?
> 
> error:
> cvOut = cvNet.forward()
> cv2.error: OpenCV(4.1.1) C:\projects\opencv-python\opencv\modules\dnn\src\layers\reshape_layer.cpp:146: error: (-1:Backtrace) Can't infer a dim denoted by -1 in function 'cv::dnn::computeShapeByReshapeMask'

I have a workaround:
    when you export your ONNX or any other formats, don't set dynamic batch_size, use a specific number.

**Comment by LupusSanctus** [comment]
Judging by ONNX diagnostics output, it seems that there will be plenty of problems during this model loading with DNN API, for example:
```
[ERROR:0] global .../opencv/modules/dnn/src/onnx/onnx_importer.cpp (486) populateNet DNN/ONNX: NOTE: Potential problem with creating node 812 with type NonZero.
 Type NonZero IS NOT SUPPORTED!

[ERROR:0] global .../opencv/modules/dnn/src/onnx/onnx_importer.cpp (486) populateNet DNN/ONNX: NOTE: Potential problem with creating node 906 with type ReduceMin.
 Type ReduceMin IS NOT SUPPORTED!

[ERROR:0] global .../opencv/modules/dnn/src/onnx/onnx_importer.cpp (486) populateNet DNN/ONNX: NOTE: Potential problem with creating node 909 with type TopK.
 Type TopK IS NOT SUPPORTED!

[ERROR:0] global .../opencv/modules/dnn/src/onnx/onnx_importer.cpp (486) populateNet DNN/ONNX: NOTE: Potential problem with creating node 1051 with type Exp.
 Type Exp IS NOT SUPPORTED!

[ERROR:0] global .../opencv/modules/dnn/src/onnx/onnx_importer.cpp (486) populateNet DNN/ONNX: NOTE: Potential problem with creating node 1059 with type Exp.
 Type Exp IS NOT SUPPORTED!

[ERROR:0] global .../opencv/modules/dnn/src/onnx/onnx_importer.cpp (486) populateNet DNN/ONNX: NOTE: Potential problem with creating node 1146 with type Less.
 Type Less IS NOT SUPPORTED!

[ERROR:0] global .../opencv/modules/dnn/src/onnx/onnx_importer.cpp (486) populateNet DNN/ONNX: NOTE: Potential problem with creating node 1147 with type Not.
 Type Not IS NOT SUPPORTED!

[ERROR:0] global .../opencv/modules/dnn/src/onnx/onnx_importer.cpp (486) populateNet DNN/ONNX: NOTE: Potential problem with creating node 1150 with type Less.
 Type Less IS NOT SUPPORTED!

[ERROR:0] global .../opencv/modules/dnn/src/onnx/onnx_importer.cpp (486) populateNet DNN/ONNX: NOTE: Potential problem with creating node 1151 with type Not.
 Type Not IS NOT SUPPORTED!

[ERROR:0] global .../opencv/modules/dnn/src/onnx/onnx_importer.cpp (486) populateNet DNN/ONNX: NOTE: Potential problem with creating node 1155 with type And.
 Type And IS NOT SUPPORTED!

[ERROR:0] global .../opencv/modules/dnn/src/onnx/onnx_importer.cpp (486) populateNet DNN/ONNX: NOTE: Potential problem with creating node 1158 with type NonZero.
 Type NonZero IS NOT SUPPORTED!

[ERROR:0] global .../opencv/modules/dnn/src/onnx/onnx_importer.cpp (486) populateNet DNN/ONNX: NOTE: Potential problem with creating node 1170 with type NonMaxSuppression.
 Type NonMaxSuppression IS NOT SUPPORTED!

[ERROR:0] global .../opencv/modules/dnn/src/onnx/onnx_importer.cpp (486) populateNet DNN/ONNX: NOTE: Potential problem with creating node 1239 with type ReduceMin.
 Type ReduceMin IS NOT SUPPORTED!
```
For running diagnostics use the below command:
```
./bin/opencv_model_diagnostics -m=./mask_rcnn_R_50_FPN_1x.onnx
```
